src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -83,8 +83,16 @@ public override void CopyTo(int sourceIndex, char[] destination, int destination
         public override SourceText WithChanges(IEnumerable<TextChange> changes)
         {
             // compute changes against newText to avoid capturing strong references to this ChangedText instance.
-            var changed = (ChangedText)_newText.WithChanges(changes);
-            return new ChangedText(this, changed._newText, changed._changes);
+            // note: _newText is never a ChangeText, so if WithChanges is a no-op the result will not be a ChangedText
+            var changed = _newText.WithChanges(changes) as ChangedText;  
+            if (changed != null)
+            {
+                return new ChangedText(this, changed._newText, changed._changes);
+            }
+            else 
+            {
+                return this;
```

Comment:
As for SourceText implementations, there are only a few types of SourceText that end up in _newText, CompositeText, SubText, LargeText or StringText.   Though it does seem possible that if the change has only one segment and that segment is the entirety of the original text, then _newText will be the original text.  So if that original text is a special subtype of SourceText that overrides the WithChanges API, it could return something that is not a ChangedText instance.